File: java/src/test/java/dev/openfeature/flagd/evaluator/comparison/MinimalInProcessResolver.java
package dev.openfeature.flagd.evaluator.comparison;

import com.fasterxml.jackson.databind.ObjectMapper;
import dev.openfeature.sdk.*;
import dev.openfeature.sdk.exceptions.GeneralError;
import dev.openfeature.sdk.exceptions.ParseError;
import dev.openfeature.sdk.exceptions.TypeMismatchError;

import java.util.HashMap;
import java.util.Map;

/**
 * Minimal in-process resolver extracted from the old java-sdk-contrib implementation.
 *
 * <p>This resolver contains ONLY the core evaluation logic without any of the infrastructure:
 * - No FlagStore/Storage layer
 * - No QueueSource/Sync connectors
 * - No event handling threads
 * - No state management
 *
 * <p>Used for comparison testing to validate that the new WASM-based evaluator
 * produces the same results as the old JsonLogic-based approach.
 *
 * <p>Source: dev.openfeature.contrib.providers.flagd.resolver.process.InProcessResolver
 */
public class MinimalInProcessResolver {

    private static final String EMPTY_TARGETING_STRING = "{}";
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final Map<String, FeatureFlag> flags = new HashMap<>();
    private final SimpleJsonLogicOperator operator = new SimpleJsonLogicOperator();

    /**
     * Manually add a flag to the resolver (replaces FlagStore).
     */
    public void setFlag(String key, FeatureFlag flag) {
        flags.put(key, flag);
    }

    /**
     * Load flags from JSON configuration (similar to updateState).
     */
    public void loadFlags(String jsonConfig) throws Exception {
        @SuppressWarnings("unchecked")
        Map<String, Object> config = MAPPER.readValue(jsonConfig, Map.class);

        @SuppressWarnings("unchecked")
        Map<String, Object> flagsMap = (Map<String, Object>) config.get("flags");

        if (flagsMap != null) {
            for (Map.Entry<String, Object> entry : flagsMap.entrySet()) {
                String flagKey = entry.getKey();
                @SuppressWarnings("unchecked")
                Map<String, Object> flagData = (Map<String, Object>) entry.getValue();

                @SuppressWarnings("unchecked")
                Map<String, Object> variants = (Map<String, Object>) flagData.get("variants");

                // Handle targeting - if not present or null, use empty object
                Object targeting = flagData.get("targeting");
                String targetingJson = (targeting == null) ? null : MAPPER.writeValueAsString(targeting);

                FeatureFlag flag = new FeatureFlag(
                    (String) flagData.get("state"),
                    (String) flagData.get("defaultVariant"),
                    variants,
                    targetingJson
                );

                setFlag(flagKey, flag);
            }
        }
    }

    /**
     * Resolve a boolean flag.
     */
    public ProviderEvaluation<Boolean> booleanEvaluation(String key, Boolean defaultValue, EvaluationContext ctx) {
        return resolve(Boolean.class, key, ctx);
    }

    /**
     * Resolve a string flag.
     */
    public ProviderEvaluation<String> stringEvaluation(String key, String defaultValue, EvaluationContext ctx) {
        return resolve(String.class, key, ctx);
    }

    /**
     * Resolve a double flag.
     */
    public ProviderEvaluation<Double> doubleEvaluation(String key, Double defaultValue, EvaluationContext ctx) {
        return resolve(Double.class, key, ctx);
    }

    /**
     * Resolve an integer flag.
     */
    public ProviderEvaluation<Integer> integerEvaluation(String key, Integer defaultValue, EvaluationContext ctx) {
        return resolve(Integer.class, key, ctx);
    }

    /**
     * Resolve an object flag.
     */
    public ProviderEvaluation<Value> objectEvaluation(String key, Value defaultValue, EvaluationContext ctx) {
        final ProviderEvaluation<Object> evaluation = resolve(Object.class, key, ctx);

        return ProviderEvaluation.<Value>builder()
                .value(Value.objectToValue(evaluation.getValue()))
                .variant(evaluation.getVariant())
                .reason(evaluation.getReason())
                .errorCode(evaluation.getErrorCode())
                .errorMessage(evaluation.getErrorMessage())
                .flagMetadata(evaluation.getFlagMetadata())
                .build();
    }

    /**
     * Core resolution logic extracted from InProcessResolver (lines 175-255).
     * Uses simple HashMap instead of FlagStore.
     */
    private <T> ProviderEvaluation<T> resolve(Class<T> type, String key, EvaluationContext ctx) {
        final FeatureFlag flag = flags.get(key);

        // missing flag
        if (flag == null) {
            return ProviderEvaluation.<T>builder()
                    .errorMessage("flag: " + key + " not found")
                    .errorCode(ErrorCode.FLAG_NOT_FOUND)
                    .flagMetadata(ImmutableMetadata.builder().build())
                    .build();
        }

        // state check
        if ("DISABLED".equals(flag.getState())) {
            return ProviderEvaluation.<T>builder()
                    .errorMessage("flag: " + key + " is disabled")
                    .errorCode(ErrorCode.FLAG_NOT_FOUND)
                    .flagMetadata(ImmutableMetadata.builder().build())
                    .build();
        }

        final String resolvedVariant;
        final String reason;

        if (EMPTY_TARGETING_STRING.equals(flag.getTargeting())) {
            resolvedVariant = flag.getDefaultVariant();
            reason = Reason.STATIC.toString();
        } else {
            try {
                final Object jsonResolved = operator.apply(key, flag.getTargeting(), ctx);
                if (jsonResolved == null) {
                    resolvedVariant = flag.getDefaultVariant();
                    reason = Reason.DEFAULT.toString();
                } else {
                    resolvedVariant = jsonResolved.toString(); // convert to string to support shorthand
                    reason = Reason.TARGETING_MATCH.toString();
                }
            } catch (Exception e) {
                String message = String.format("error evaluating targeting rule for flag %s", key);
                throw new ParseError(message);
            }
        }

        // check variant existence
        Object value = flag.getVariants().get(resolvedVariant);
        if (value == null) {
            String message = String.format("variant %s not found in flag with key %s", resolvedVariant, key);
            throw new GeneralError(message);
        }

        if (value instanceof Integer && type == Double.class) {
            // if this is an integer and we are trying to resolve a double, convert
            value = ((Integer) value).doubleValue();
        } else if (value instanceof Double && type == Integer.class) {
            // if this is a double and we are trying to resolve an integer, convert
            value = ((Double) value).intValue();
        }

        if (!type.isAssignableFrom(value.getClass())) {
            String message = "returning default variant for flagKey: %s, type not valid";
            throw new TypeMismatchError(message);
        }

        return ProviderEvaluation.<T>builder()
                .value((T) value)
                .variant(resolvedVariant)
                .reason(reason)
                .flagMetadata(ImmutableMetadata.builder().build())
                .build();
    }

    /**
     * Simple feature flag model (extracted from old implementation).
     */
    public static class FeatureFlag {
        private final String state;
        private final String defaultVariant;
        private final Map<String, Object> variants;
        private final String targeting;

        public FeatureFlag(String state, String defaultVariant,
                          Map<String, Object> variants, String targeting) {
            this.state = state;
            this.defaultVariant = defaultVariant;
            this.variants = variants;
            this.targeting = targeting;
        }

        public String getState() {
            return state;
        }

        public String getDefaultVariant() {
            return defaultVariant;
        }

        public Map<String, Object> getVariants() {
            return variants;
        }

        public String getTargeting() {
            return targeting == null ? EMPTY_TARGETING_STRING : targeting;
        }
    }
}

File: dotnet/benchmarks/FlagdEvaluator.Benchmarks/MinimalInProcessResolver.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using Json.Logic;

namespace FlagdEvaluator.Benchmarks;

/// <summary>
/// Minimal in-process resolver extracted from the old dotnet-sdk-contrib flagd provider.
/// Uses json-everything's JsonLogic for targeting evaluation — the same library as the
/// production OpenFeature.Contrib.Providers.Flagd package.
///
/// Contains ONLY the core evaluation path for fair benchmarking:
/// 1. Parse targeting rule from string → JsonNode (per evaluation, like the old provider)
/// 2. Serialize context dictionary → JSON string → JsonNode (per evaluation)
/// 3. JsonLogic.Apply(rule, data) → variant name
/// 4. Variant lookup → value
///
/// No FlagStore, no sync connectors, no event handling.
/// </summary>
internal sealed class MinimalInProcessResolver
{
    private readonly Dictionary<string, FeatureFlag> _flags = new();

    internal void LoadFlags(string jsonConfig)
    {
        using var doc = JsonDocument.Parse(jsonConfig);
        var flagsObj = doc.RootElement.GetProperty("flags");

        foreach (var flagProp in flagsObj.EnumerateObject())
        {
            var flagKey = flagProp.Name;
            var flagData = flagProp.Value;

            var state = flagData.GetProperty("state").GetString()!;
            var defaultVariant = flagData.GetProperty("defaultVariant").GetString()!;

            var variants = new Dictionary<string, JsonElement>();
            foreach (var v in flagData.GetProperty("variants").EnumerateObject())
            {
                variants[v.Name] = v.Value.Clone();
            }

            string? targeting = null;
            if (flagData.TryGetProperty("targeting", out var targetingEl) &&
                targetingEl.ValueKind != JsonValueKind.Null)
            {
                targeting = targetingEl.GetRawText();
            }

            _flags[flagKey] = new FeatureFlag(state, defaultVariant, variants, targeting);
        }
    }

    /// <summary>
    /// Evaluate a flag following the old provider's hot path.
    /// This intentionally re-parses the targeting rule and re-serializes context
    /// on every call, matching the old provider's behavior.
    /// </summary>
    internal OldEvaluationResult Evaluate(string flagKey, Dictionary<string, object?>? context)
    {
        if (!_flags.TryGetValue(flagKey, out var flag))
        {
            return new OldEvaluationResult
            {
                Reason = "FLAG_NOT_FOUND",
                ErrorCode = "FLAG_NOT_FOUND",
                ErrorMessage = $"flag: {flagKey} not found",
            };
        }

        if (flag.State == "DISABLED")
        {
            return new OldEvaluationResult
            {
                Reason = "DISABLED",
                ErrorCode = "FLAG_NOT_FOUND",
                ErrorMessage = $"flag: {flagKey} is disabled",
            };
        }

        string resolvedVariant;
        string reason;

        if (flag.Targeting == null)
        {
            resolvedVariant = flag.DefaultVariant;
            reason = "STATIC";
        }
        else
        {
            // Step 1: Parse targeting rule from string (per-evaluation, like old provider)
            var rule = JsonNode.Parse(flag.Targeting);

            // Step 2: Serialize context to JSON string then parse to JsonNode
            // This mirrors the old provider's ConvertToDynamicObject → Serialize → Parse path
            JsonNode? data;
            if (context != null && context.Count > 0)
            {
                var contextJson = JsonSerializer.Serialize(context);
                data = JsonNode.Parse(contextJson);
            }
            else
            {
                data = JsonNode.Parse("{}");
            }

            // Step 3: Evaluate using json-everything's JsonLogic
            var result = JsonLogic.Apply(rule, data);

            if (result == null)
            {
                resolvedVariant = flag.DefaultVariant;
                reason = "DEFAULT";
            }
            else
            {
                resolvedVariant = result.ToString();
                reason = "TARGETING_MATCH";
            }
        }

        if (!flag.Variants.TryGetValue(resolvedVariant, out var value))
        {
            return new OldEvaluationResult
            {
                Reason = "ERROR",
                ErrorCode = "GENERAL",
                ErrorMessage = $"variant {resolvedVariant} not found in flag {flagKey}",
            };
        }

        return new OldEvaluationResult
        {
            Value = value,
            Variant = resolvedVariant,
            Reason = reason,
        };
    }

    internal sealed class FeatureFlag
    {
        public string State { get; }
        public string DefaultVariant { get; }
        public Dictionary<string, JsonElement> Variants { get; }
        public string? Targeting { get; }

        public FeatureFlag(string state, string defaultVariant,
            Dictionary<string, JsonElement> variants, string? targeting)
        {
            State = state;
            DefaultVariant = defaultVariant;
            Variants = variants;
            Targeting = targeting;
        }
    }
}

public sealed class OldEvaluationResult
{
    public JsonElement? Value { get; init; }
    public string? Variant { get; init; }
    public string? Reason { get; init; }
    public string? ErrorCode { get; init; }
    public string? ErrorMessage { get; init; }
}
